uaclient/actions.py

```diff
@@ -117,3 +119,62 @@ def status(
         ret = 0
 
     return status, ret
+
+
+def get_cloud_instance(
+    cfg: config.UAConfig,
+) -> Optional[AutoAttachCloudInstance]:
+    disable_auto_attach = util.is_config_value_true(
+        config=cfg.cfg, path_to_value="features.disable_auto_attach"
+    )
+    if disable_auto_attach:
+        msg = "Skipping auto-attach. Config disable_auto_attach is set."
+        logging.debug(msg)
+        print(msg)
+        return None
+
+    instance = None  # type: Optional[AutoAttachCloudInstance]
+    try:
+        instance = identity.cloud_instance_factory()
+    except exceptions.CloudFactoryError as e:
+        if cfg.is_attached:
+            # We are attached on non-Pro Image, just report already attached
+            raise exceptions.AlreadyAttachedError(cfg)
+        if isinstance(e, exceptions.CloudFactoryNoCloudError):
+            raise exceptions.UserFacingError(
+                messages.UNABLE_TO_DETERMINE_CLOUD_TYPE
```

Comment:
Let's add `msg_code` to all of those exceptions

Reply:
Alright.